src/System.Private.CoreLib/shared/System/Text/ASCIIEncoding.cs

```diff
@@ -357,564 +398,511 @@ public override unsafe string GetString(byte[] bytes, int byteIndex, int byteCou
+            // If we got this far, we have complex fallback semantics.
+            // Let's check that we really need to invoke the fallback before we go down the error handling path.
 
-                // We mustn't have left over fallback data when counting
-                if (encoder.InternalHasFallbackBuffer)
-                {
-                    // We always need the fallback buffer in get bytes so we can flush any remaining ones if necessary
-                    fallbackBuffer = encoder.FallbackBuffer;
-                    if (fallbackBuffer.Remaining > 0 && encoder._throwOnOverflow)
-                        throw new ArgumentException(SR.Format(SR.Argument_EncoderFallbackNotEmpty, this.EncodingName, encoder.Fallback.GetType()));
+            indexOfFirstNonAsciiChar = (int)AsciiUtility.GetIndexOfFirstNonAsciiChar(pChars, (uint)charCount);
+            Debug.Assert(0 <= indexOfFirstNonAsciiChar && indexOfFirstNonAsciiChar <= charCount);
 
-                    // Set our internal fallback interesting things.
-                    fallbackBuffer.InternalInitialize(chars, charEnd, encoder, false);
-                }
-
-                // Verify that we have no fallbackbuffer, for ASCII its always empty, so just assert
-                Debug.Assert(!encoder._throwOnOverflow || !encoder.InternalHasFallbackBuffer ||
-                    encoder.FallbackBuffer.Remaining == 0,
-                    "[ASCIICodePageEncoding.GetByteCount]Expected empty fallback buffer");
-            }
-            else
+            if (indexOfFirstNonAsciiChar != charCount)
             {
-                fallback = this.EncoderFallback as EncoderReplacementFallback;
+                goto InvokeFallback; // found non-ASCII data and couldn't continue
             }
 
-            // If we have an encoder AND we aren't using default fallback,
-            // then we may have a complicated count.
-            if (fallback != null && fallback.MaxCharCount == 1)
-            {
-                // Replacement fallback encodes surrogate pairs as two ?? (or two whatever), so return size is always
-                // same as input size.
-                // Note that no existing SBCS code pages map code points to supplimentary characters, so this is easy.
+        Finished:
 
-                // We could however have 1 extra byte if the last call had an encoder and a funky fallback and
-                // if we don't use the funky fallback this time.
+            return indexOfFirstNonAsciiChar;
 
-                // Do we have an extra char left over from last time?
-                if (charLeftOver > 0)
-                    charCount++;
+        InvokeFallback:
 
-                return (charCount);
-            }
+            // Otherwise, will need to invoke the fallback handler.
+            indexOfFirstNonAsciiChar = GetByteCount_Fallback(new ReadOnlySpan<char>(pChars, charCount), encoder, indexOfFirstNonAsciiChar);
```

Comment:
Should the fallback take pointers, so that the fast path here is as lean as possible?

Reply:
You mean using a trampoline like in the `GetBytes` and `GetChars` methods? I had this originally, but it turns out the stack spillage of a single Span didn't show up on perf runs, so I removed it for the `GetByteCount` / `GetCharCount` methods. We can definitely add it back in for consistency though if you think it's beneficial.

Reply:
No, just make the fallback methods take pointers.

Reply:
And the fallback method itself can make the Span out of the pointers as the first thing.